README.pod
=head1 NAME

Tusu - Apache-like dispatcher for Mojolicious

=head1 SYNOPSIS

    use Tusu;

For non lite app

    sub startup {
        my $self = shift;
        my $tusu = $self->plugin(tusu => {});
    }

OR

    sub startup {
        my $self = shift;
        my $tusu = $self->plugin(tusu => {
            document_root => $self->home->rel_dir('www2'),
            components => {
                'Your::Component' => 'YC',
            },
            extensions_to_render => [qw(html htm xml txt)],
        });
        
        $r->route('/specific/path')->to(cb => sub {
            $tusu->bootstrap($_[0], 'Your::Component', 'your_method');
        });
    }

For lite app

    my $tusu = plugin tusu => {...};

=head1 DESCRIPTION

Note that this plugin is formerly developed as MojoX::Tusu.

C<Tusu> is a sub framework on Mojolicious using
Text::PSTemplate for renderer. With this framework, you can deploy directory
based web sites onto Mojolicious at once.

This framework automatically activate own dispatcher which behaves like apache
web server. You can build your web site into single document root directory
named public_html in hierarchal structure. The document root directory can
contain both server-parsed-documents and static files such as images.

Tusu doesn't require files to be named like index.html.ep
style but just like index.html. You can specify which files to be server
parsable by telling it the extensions. It also provides some more apache-like
features such as directory_index, error_document and file permissions checking.

One of the intent of this module is to enhance existing static websites into
dynamic with minimal effort. The chances are that most typical website data are
transplantable with no change at all.

=head2 Installation

    $ sudo -s 'curl -L cpanmin.us | perl - https://github.com/jamadam/Tusu/tarball/master/v0.25'

=head2 Getting Started

    $ mojo generate tusu_app MyApp
    $ cd ./my_app
    $ prove
    $ ./script/my_app daemon
    Server available at http://127.0.0.1:3000.

=head2 Template Syntax

See L<https://github.com/jamadam/Text-PSTemplate> for detail.

In addition to Text::PSTemplate's default syntax, Tusu
provides short cut for html escaping as follows

    <% $var %> normal
    <%= $var %> escaped
    <%= some_func(...) %> escaped

=head2 Components

Mojo::Tusu provides object oriented component framework. You can easily add your
custom features into your website. The following is an example for component
development.

    <span><% questionize('Hello') %></span>

To make it possible, you should write a module like this. 

    package MyUtility;
    use strict;
    use warnings;
    use base 'Tusu::ComponentBase';
    
    sub questionize : TplExport {
        my ($self, $sentence) = @_;
        my $c = $self->controller; # mojolicious controller in case you need
        return $sentence . '?';
    }

To activate this component, you must plug-in this at mojolicious startup method.

    sub startup {
        my $self = shift;
        my $tusu = $self->plugin(tusu => {
            components => {
                'YourUtility' =>  '' ## namespace is ''
            }
        });
    }

Here is another example for component development.

    <div id="productContainer">
        <% Product::list_by_category('books', 10) %>
    </div>

To make it possible, you should write a module like this.

    package Product;
    use strict;
    use warnings;
    use base 'Tusu::ComponentBase';
    
    __PACKAGE__->attr('some_data');
    
    sub init {
        my ($self, $app) = @_;
        $self->some_data('value'); ### DB SETTING OR SOMETHING
    }
    
    sub list_by_category : TplExport {
        my ($self, $category, $limit) = @_;
        my $c = $self->controller; # mojolicious controller in case you need
        
        # MAY BE ACCESS TO YOUR DB HERE
        
        return $html_snippet;
    }

To activate this component, you must plug-in this at mojolicious startup method.

    sub startup {
        my $self = shift;
        my $tusu = $self->plugin(tusu => {
            components => {
                Product => undef
            },
        });
    }

=head1 OPTIONS

=head2 document_root => string

This option sets root directory for templates and static files. Following
example is default setting.

    my $tusu = $self->plugin(tusu => {
        document_root => $self->home->rel_dir('public_html')
    });

=head2 components => hash

    my $tusu = $self->plugin(tusu => {
        components => {
            'Namespace::A' => 'A',   # namespace is A
            'Namespace::B' => '',    # namespace is ''
            'Namespace::C' => undef, # namespace is Namespace::C
        },
    });

=head2 encoding => string or array ref

This option sets encoding for template files. Array ref causes auto detection
active.

    my $tusu = $self->plugin(tusu => {
        encoding => 'Shift-JIS',
    });
    
    or..
    
    my $tusu = $self->plugin(tusu => {
        encoding => ['Shift-JIS', 'utf8'],
    });

=head2 directory_index => array ref

This option sets default file names for searching files in directory when
the request path doesn't ended with file name. And this setting also affects to
inside template context such as include('path') function. Following example is
the default setting.

    my $tusu = $self->plugin(tusu => {
        directory_index => ['index.html', 'index.htm'],
    });

=head2 indexes => bool

This option emulates apache's indexes option. When the value is 1,
the server generates file list page for directory access.

    my $tusu = $self->plugin(tusu => {
        indexes => 1,
    });

=head2 extensions_to_render => array ref

This option sets the extensions to be parsed by tusu renderer. If request
doesn't match any of extensions, dispatcher try to render it as static file.
Following setting is the default.

    my $tusu = $self->plugin(tusu => {
        extensions_to_render => ['html','htm','xml'],
    });

=head2 error_document => hash ref

This option setup custom error pages like apache's ErrorDocument.

    my $tusu = $self->plugin(tusu => {
        error_document => {
            404 => '/errors/404.html',
            403 => '/errors/403.html',
            500 => '/errors/405.html',
        },
    });

=head1 METHODS

=head2 Tusu->new($app)

Constructor. 
    
    $tusu = Tusu->new($app)

=head2 $instance->register($app)

This method internally called.

=head2 $instance->engine

This returns Text::PSTemplate instance. You can customize the template system
behavior by calling parser methods directly.
    
    my $tusu = Tusu->new($app);
    my $pst = $tusu->engine;
    $pst->set_delimiter('<!--', '-->');

=head2 $instance->bootstrap($controller, $component, $method)

This method is a sub dispatcher method. You can specify a class and a method the
route to be dispatched to.

    my $tusu = Tusu->new($self);
    $r->route('/some/path')->via('post')->to(cb => sub {
        $tusu->bootstrap($c, 'Your::Component', 'post');
    });

=head2 $instance->get_component($name)

This is an alias to Text::PSTemplate->get_plugin. With this method, you can get
component instance.

    my $component = $tusu->get_component('Your::Component');

=head1 What does Tusu mean?

Tusu means mojo in Ainu languages which is spoken by the native inhabitants of
Hokkaido prefecture, Japan.

=head1 SEE ALSO

L<Mojolicious>, L<Text::PSTemplate>

L<http://en.wikipedia.org/wiki/Ainu_languages>

=head1 AUTHOR

Sugama Keita, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Sugama Keita.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


Changes
Revision history for Tusu

0.02 2012/03/07
  - Upgraded Text::PSTemplate to v0.47
  - Fixed version number

0.01 2012/02/25
  - Initial release 


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

my %params = (
    NAME          => 'Tusu',
    AUTHOR        => 'sugama <sugama@jamadam.com>',
    VERSION_FROM  => 'lib/Tusu.pm',
    ABSTRACT_FROM => 'lib/Tusu.pm',
    LICENSE       => 'perl',
    PREREQ_PM     => {
    },
    BUILD_REQUIRES => {
        'Test::More'          => '0.47',
        'Test::UseAllModules' => '0.10',
        'Mojolicious'    	  => '2.52',
        'Attribute::Handlers' => '0.91',
        'Try::Tiny'           => '0.11',
        'Carp'                => '1.25',
    },
    META_MERGE => {
        resources => {
            repository => '',
        },
    },
	test => {TESTS => 't/*.t t/*/*.t'},
);

my $eumm = $ExtUtils::MakeMaker::VERSION;
delete $params{LICENSE}          if $eumm < 6.31;
delete $params{MIN_PERL_VERSION} if $eumm < 6.48;
delete $params{META_MERGE}       if $eumm < 6.46;
delete $params{META_ADD}         if $eumm < 6.46;
delete $params{LICENSE}          if $eumm < 6.31;

if ($eumm < 6.52 && $params{CONFIGURE_REQUIRES}) {
    $params{PREREQ_PM} = {
        %{ $params{PREREQ_PM}          || {} },
        %{ $params{CONFIGURE_REQUIRES} },
    };
    delete $params{CONFIGURE_REQUIRES};
}
if ($eumm < 6.5503 && $params{BUILD_REQUIRES}) {
    $params{PREREQ_PM} = {
        %{ $params{PREREQ_PM}      || {} },
        %{ $params{BUILD_REQUIRES} },
    };
    delete $params{BUILD_REQUIRES};
}

WriteMakefile(%params);


lib/Tusu/ComponentBase.pm
package Tusu::ComponentBase;
use strict;
use warnings;
use Mojo::Base;
use Tusu::UserError;
use base qw(Text::PSTemplate::PluginBase);

    sub attr {
        Mojo::Base::attr(@_);
    }

    sub new {
        my $class = shift;
        my $self = $class->SUPER::new(@_);
        $self->init($Tusu::APP);
        return $self;
    }
    
    sub init {
        ### Must be implemented on sub classes.
    }
	
	sub component : TplExport {
		my ($self, $name) = @_;
		if ($name) {
			return $self->get_engine->get_plugin($name);
		}
		return $self;
	}
    
    sub _dummy : TplExport {
        
    }
    
    sub controller {
        return $Tusu::CONTROLLER;
    }
    
    sub render {
        my ($self) = shift;
        my $c = $Tusu::CONTROLLER;
        $c->render(
            handler => 'tusu',
            template => $c->req->url->path->to_string,
            @_
        );
    }
	
	sub redirect_to {
		my ($self, $url) = @_;
		my $c = $self->controller;
		
		my $base;
		if ($ENV{REQUEST_URI}) {
            my $path = $ENV{REQUEST_URI};
            $path =~ s{\?.+}{};
			$base = $c->req->url->base->clone->path($path)->to_abs;
		} else {
			$base = $c->req->url->clone->to_abs;
		}
		$base->userinfo(undef);
		
		my $res     = $c->res;
		my $headers = $res->headers;
		$headers->location(Mojo::URL->new($url)->base($base)->to_abs);
		$c->rendered($res->is_status_class(300) ? undef : 302);
		return $c;
	}
	
	### ---
	### user_error
	### ---
    sub user_err : TplExport {
        my ($self) = @_;
        my $c = $self->controller;
        if (! $c->stash('user_err')) {
            $c->stash('user_err', Tusu::UserError->new)
        }
        return $c->stash('user_err');
    }

1;

__END__

=head1 NAME

Tusu::ComponentBase - Base Class for WAF component

=head1 SYNOPSIS
    
    package YourComponent;
    use strict;
    use warnings;
    use base qw(Tusu::ComponentBase);
    
    sub your_action1 {
        my ($self, $controller) = @_;
        $controller->render(
            handler => 'tusu',
            format  => 'html',
            template => 'some_template',
        );
    }
    sub your_action2 {
        my ($self, $controller) = @_;
        # ...
    }
    # ...
    sub some_func : TplExport {
        my ($self, @your_args) = @_;
        # ...
        return '';
    }
    
    <% YourComponent::some_func(@your_args) %>
    <% YourComponent::component() %>

=head1 DESCRIPTION

C<Tusu::ComponentBase> is a Component Base class for
Tusu sub framework on mojolicious. This class inherits
all methods from Text::PSTemplate::PluginBase.

=head1 METHODS

=head2 controller

Returns current Mojolicious::Controller instance.

=head2 $self->init($app)

This is a hook method for initializing component. This will automatically be
called from constructor.

=head2 $instance->attr

=head2 $class->new

=head2 $instance->redirect_to

This is a wrapper method for $c->redirect_to to avoid using PATH_INFO on CGI
environment.

    $self->redirect_to('./foo.html');
    $self->redirect_to('/foo/bar.html');
    $self->redirect_to('http://example.com/foo/bar.html');
    $self->redirect_to('/');

=head2 $instance->component($class)

Returns component instance of given class. If $class is null, this returns
current class instance.

=head2 $instance->user_error

=head1 SEE ALSO

L<Text::PSTemplate>, L<Mojolicious::Plugin::Renderer>

=head1 AUTHOR

Sugama Keita, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Sugama Keita.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


lib/Mojolicious/Plugin/TusuRenderer.pm
package Mojolicious::Plugin::TusuRenderer;
use strict;
use warnings;
use Try::Tiny;
use Text::PSTemplate;
use Mojo::Base 'Tusu';
use Carp;
use Scalar::Util qw(weaken);

    __PACKAGE__->attr('handler' => 'tusu');
    __PACKAGE__->attr('engine');
    
    sub default_args {
        my $class = shift;
        return {
            %{$class->SUPER::default_args(@_)},
            encoding                => 'utf8',
            components              => {},
        }
    }
    
    sub register {
        my ($self, $app, $args) = @_;
        
        $self->SUPER::register($app, $args);
        
        my $engine = Text::PSTemplate->new;
        $engine->set_filter('=', \&Mojo::Util::html_escape);
        $engine->set_filename_trans_coderef(sub {
            $self->filename_trans($app->static->paths->[0], @_);
        });
        
        {
            local $Tusu::APP = $app;
            $engine->plug(
                'Tusu::ComponentBase'            => undef,
                'Tusu::Component::Util'          => '',
                'Tusu::Component::Mojolicious'   => 'Mojolicious',
                %{$args->{components}}
            );
        }
        $self->engine($engine);
        $engine->set_encoding($args->{encoding});
        
        $app->renderer->add_handler(tusu => sub {
            my ($renderer, $c, $output, $options) = @_;
            
            try {
                local $SIG{__DIE__} = undef;
                local $Tusu::CONTROLLER = $c;
                $$output = Text::PSTemplate ->new($self->engine)
                                            ->parse_file('/'. $options->{template});
            } catch {
                my $err = $_ || 'Unknown Error';
                $c->app->log->error(qq(Template error in "$options->{template}": $err));
                $c->render_exception("$err");
                $$output = '';
                return 0;
            };
            return 1;
        });
        
        weaken $self;
        
        return $self;
    }
    
    ### ---
    ### Get component
    ### ---
    sub get_component {
        my ($self, $name) = @_;
        return $self->engine->get_plugin($name);
    }
    
    ### ---
    ### bootstrap for frameworking
    ### ---
    sub bootstrap {
        my ($self, $c, $component, $action, @args) = @_;
        local $Tusu::CONTROLLER = $c;
        return $self->engine->get_plugin($component)->$action(@args);
    }

1;

__END__

=head1 NAME

Tusu::TusuRenderer - Tusu::TusuRenderer

=head1 SYNOPSIS

=head1 DESCRIPTION

=head1 METHODS

=head1 SEE ALSO

L<Text::PSTemplate>, L<Mojolicious::Plugin::Renderer>

=head1 AUTHOR

Sugama Keita, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Sugama Keita.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


lib/Tusu.pm
package Tusu;
use strict;
use warnings;
use Try::Tiny;
use Text::PSTemplate;
use Mojo::Base 'Mojolicious::Plugin';
use Mojo::Util;
use Carp;
use File::Basename;
use File::Spec;
use Mojolicious::Types;
use Mojo::Util 'url_unescape';
use Scalar::Util qw(weaken);
our $VERSION = '0.02';

    our $APP;
    our $CONTROLLER;
    
    __PACKAGE__->attr('extensions_to_render');
    __PACKAGE__->attr('directory_index');
    __PACKAGE__->attr('error_document');
    __PACKAGE__->attr('document_root');
    __PACKAGE__->attr('indexes');
    
    sub default_args {
        my ($class, $app) = @_;
        return {
            document_root           => $app->home->rel_dir('public_html'),
            extensions_to_render    => ['html','htm','xml'],
            directory_index         => ['index.html','index.htm'],
            error_document          => {},
            indexes                 => 0,
        }
    }
    
    sub _check_hash_key {
        my ($hash_ref, @keys) = @_;
        my %keys;
        $keys{$_} = 1 foreach(@keys);
        for my $key (keys %$hash_ref) {
            if (! $keys{$key}) {
                return $key;
            }
        }
    }

    sub register {
        my ($self, $app, $args) = @_;
        
        $args = {%{$self->default_args($app)}, %$args};
        
        if (my $key = _check_hash_key($args, keys %$args)) {
            croak "Unknown argument $key";
        }
        
        if (! -d $args->{document_root}) {
            die qq{Document root "$args->{document_root}" is not a directory};
        }
        
        my $tusu_default_route_set;
        
        $app->hook('around_dispatch' => sub {
            my $app = $_[1]->app;
            if (! $tusu_default_route_set) {
                $tusu_default_route_set = 1;
                $_[1]->app->routes
                    ->route('/:template', template => qr{.*})
                    ->name('')
                    ->to(cb => sub {
                        $_[0]->render(handler => $self->handler);
                    });
            }
            $self->_dispatch($_[1]);
        });
        
        $app->static->paths([$args->{document_root}, $self->_asset]);
        $app->renderer->paths([$args->{document_root}]);
        
        $self->directory_index($args->{directory_index});
        $self->error_document($args->{error_document});
        $self->extensions_to_render($args->{extensions_to_render});
        $self->document_root($args->{document_root});
        $self->indexes($args->{indexes});
        
        return $self;
    }
    
    ### ---
    ### Custom dispatcher
    ### ---
    sub _dispatch {
        my ($self, $c) = @_;
        my $app = $c->app;
        
        my $tx = $c->tx;
        if ($tx->is_websocket) {
            $c->res->code(undef);
        }
        $app->sessions->load($c);
        my $plugins = $app->plugins;
        $plugins->emit_hook(before_dispatch => $c);
        
        my $path = $tx->req->url->path->to_string || '/';
        
        my $not_found;
        
        my $check_result = $self->_check_file_type($path);
        
        if (! $check_result->{type}) {
            if ($path =~ qr{/$} && $self->indexes) {
                $self->_render_indexes($c);
                return;
            } else {
                $not_found = 1;
            }
        } elsif ($check_result->{type} eq 'directory') {
            $c->redirect_to($path. '/');
            $tx->res->code(301);
            return;
        } elsif (! _permission_ok($check_result->{path}, $app->static->paths->[0])) {
            $self->_render_error_document($c, 403);
            return;
        }
        
        if (! $not_found) {
            
            my $path = $check_result->{path};
            
            ### dynamic content
            for my $ext (@{$self->extensions_to_render}) {
                if ($path !~ m{\.} || $path =~ m{\.$ext$}) {
                    my $res = $tx->res;
                    if (my $code = ($tx->req->error)[1]) {
                        $res->code($code)
                    } elsif ($tx->is_websocket) {
                        $res->code(426)
                    }
                    if (! $app->routes->dispatch($c) || ! $res->code) {
                        $c->render_not_found;
                    }
                    return;
                }
            }
            
            ## This must not be happen
            if ($path =~ m{((\.(cgi|php|rb))|/)$}) {
                $self->_render_error_document($c, 403);
                return;
            }
        }
        
        my $relpath =
            ($check_result->{path})
                ? File::Spec->abs2rel($check_result->{path}, $app->static->paths->[0])
                : $path;
        ### defaults to static content
        if ($app->static->serve($c, $relpath)) {
            $c->stash->{'mojo.static'} = 1;
            $c->rendered;
        }
        if (! $tx->res->code) {
            $self->_render_error_document($c, 404);
        }
        $plugins->emit_hook_reverse(after_static_dispatch => $c);
    }

    ### ---
    ### Asset directory
    ### ---
    sub _asset {
        my ($class, $path) = @_;
        if ($path) {
            return File::Spec->catfile(dirname(__FILE__), 'Tusu', 'Asset', $path);
        }
        return File::Spec->catdir(dirname(__FILE__), 'Tusu', 'Asset');
    }
    
    ### ---
    ### Guess type by file extension
    ### ---
    sub _file_to_mime_class {
        my $name = shift;
        my $ext = ($name =~ qr{\.(\w+)$}) ? $1 : '';
        return (split('/', Mojolicious::Types->type($ext) || 'text/plain'))[0];
    }
    
    ### ---
    ### Get file utime
    ### ---
    sub _file_timestamp {
        my $path = shift;
        my @dt = localtime((stat($path))[9]);
        return sprintf('%d-%02d-%02d %02d:%02d', 1900 + $dt[5], $dt[4] + 1, $dt[3], $dt[2], $dt[1]);
    }
    
    ### ---
    ### Get file size
    ### ---
    sub _file_size {
        my $path = shift;
        return ((stat($path))[7] > 1024)
            ? sprintf("%.1f",(stat($path))[7] / 1024) . 'KB'
            : (stat($path))[7]. 'B';
    }
    
    ### ---
    ### Render file list
    ### ---
    sub _render_indexes {
        my ($self, $c) = @_;
        my $req_path = url_unescape($c->tx->req->url->path);
        utf8::decode($req_path);
        my $dir = $self->document_root() . $req_path;
        
        opendir(my $DIR, $dir);
        my @file = readdir($DIR);
        closedir $DIR;
        
        my @dataset = ();
        for my $file (@file) {
            utf8::decode($file);
            $file = url_unescape($file);
            if ($file =~ qr{^\.$} || $req_path =~ qr{^/$} && $file =~ qr{^\.\.$}) {
                next;
            }
            my $path = $dir. $file;
            push(@dataset, {
                name        => -f $path ? $file : $file. '/',
                timestamp   => _file_timestamp($path),
                size        => _file_size($path),
                type        => -f $path ? _file_to_mime_class($file) : 'dir',
            });
        }
        @dataset = sort {
            ($a->{type} ne 'dir') <=> ($b->{type} ne 'dir')
            ||
            $a->{name} cmp $b->{name}
        } @dataset;
        my $tpl = Text::PSTemplate->new;
        $tpl->set_var(dir => $req_path, dataset => \@dataset);
        $c->render_text($tpl->parse_file($self->_asset('file_list.html')));
    }
    
    ### ---
    ### Render Error document
    ### ---
    sub _render_error_document {
        my ($self, $c, $code, $debug_message) = @_;
        
        $debug_message ||= 'Unknown Error';
        
        $c->app->log->debug($debug_message);
        
        if ($c->app->mode eq 'production') {
            if (my $template = $self->error_document->{$code}) {
                $c->render(handler => 'tusu', template => $template);
                $c->rendered($code);
                return;
            }
        }
        if ($code == 404) {
            $c->render_not_found;
        } else {
            $c->render_exception($debug_message);
        }
        $c->res->code($code);
    }
    
    ### ---
    ### fill directory_index candidate
    ### ---
    sub fill_filename {
        my ($self, $path) = @_;
        for my $default (@{$self->directory_index}) {
            my $path = File::Spec->catfile($path, $default);
            if (-f $path) {
                return $path;
            }
        }
        return;
    }
    
    ### ---
    ### find file and type
    ### ---
    sub _check_file_type {
        my ($self, $name) = @_;
        $name ||= '';
        my $leading_slash  = (substr($name, 0, 1) eq '/');
        my $trailing_slash = (substr($name, -1, 1) eq '/');
        $name =~ s{^/}{};
        my $path = File::Spec->catfile($self->document_root, $name);
        if (-f $path) {
            return {type => 'file', path => $path};
        }
        if ($trailing_slash) {
            if (my $fixed_path = $self->fill_filename($path)) {
                return {type => 'file', path => $fixed_path};
            }
        } elsif (-d $path) {
            return {type => 'directory'};
        }
        return {};
    }
    
    ### ---
    ### Check if others readable
    ### ---
    sub _permission_ok {
        my ($name, $base) = @_;
        $base ||= '';
        if ($^O eq 'MSWin32') {
            return 1;
        }
        if ($name && -f $name && ((stat($name))[2] & 4)) {
            $name =~ s{(^|/)[^/]+$}{};
            while (-d $name) {
                if ($name eq $base) {
                    return 1;
                }
                if (! ((stat($name))[2] & 1)) {
                    return 0;
                }
                $name =~ s{(^|/)[^/]+$}{};
            }
            return 1;
        }
        return 0;
    }
    
    ### ---
    ### foo/bar.html    -> public_html/foo/bar.html
    ### foo/            -> public_html/foo/index.html
    ### foo             -> public_html/foo
    ### ---
    sub filename_trans {
        my ($self, $template_base, $name) = @_;
        $name ||= '';
        my $leading_slash = substr($name, 0, 1) eq '/';
        my $trailing_slash = substr($name, -1, 1) eq '/';
        $name =~ s{^/}{};
        my $dir;
        if ($leading_slash) {
            $dir = $template_base;
        } else {
            $dir = (File::Spec->splitpath(Text::PSTemplate->get_current_filename))[1];
        }
        my $path = File::Spec->catfile($dir, $name);
        if ($trailing_slash) {
            if (my $fixed_path = $self->fill_filename($path)) {
                return $fixed_path;
            }
        }
        return $path;
    }

1;

__END__

=head1 NAME

Tusu - Apache-like dispatcher for Mojolicious

=head1 SYNOPSIS

    use Tusu;

For non lite app

    sub startup {
        my $self = shift;
        my $tusu = $self->plugin(tusu => {});
    }

OR

    sub startup {
        my $self = shift;
        my $tusu = $self->plugin(tusu => {
            document_root => $self->home->rel_dir('www2'),
            components => {
                'Your::Component' => 'YC',
            },
            extensions_to_render => [qw(html htm xml txt)],
        });
        
        $r->route('/specific/path')->to(cb => sub {
            $tusu->bootstrap($_[0], 'Your::Component', 'your_method');
        });
    }

For lite app

    my $tusu = plugin tusu => {...};

=head1 DESCRIPTION

Note that this plugin is formerly developed as MojoX::Tusu.

C<Tusu> is a sub framework on Mojolicious using
Text::PSTemplate for renderer. With this framework, you can deploy directory
based web sites onto Mojolicious at once.

This framework automatically activate own dispatcher which behaves like apache
web server. You can build your web site into single document root directory
named public_html in hierarchal structure. The document root directory can
contain both server-parsed-documents and static files such as images.

Tusu doesn't require files to be named like index.html.ep
style but just like index.html. You can specify which files to be server
parsable by telling it the extensions. It also provides some more apache-like
features such as directory_index, error_document and file permissions checking.

One of the intent of this module is to enhance existing static websites into
dynamic with minimal effort. The chances are that most typical website data are
transplantable with no change at all.

=head2 Installation

    $ sudo -s 'curl -L cpanmin.us | perl - https://github.com/jamadam/Tusu/tarball/master/v0.25'

=head2 Getting Started

    $ mojo generate tusu_app MyApp
    $ cd ./my_app
    $ prove
    $ ./script/my_app daemon
    Server available at http://127.0.0.1:3000.

=head2 Template Syntax

See L<https://github.com/jamadam/Text-PSTemplate> for detail.

In addition to Text::PSTemplate's default syntax, Tusu
provides short cut for html escaping as follows

    <% $var %> normal
    <%= $var %> escaped
    <%= some_func(...) %> escaped

=head2 Components

Mojo::Tusu provides object oriented component framework. You can easily add your
custom features into your website. The following is an example for component
development.

    <span><% questionize('Hello') %></span>

To make it possible, you should write a module like this. 

    package MyUtility;
    use strict;
    use warnings;
    use base 'Tusu::ComponentBase';
    
    sub questionize : TplExport {
        my ($self, $sentence) = @_;
        my $c = $self->controller; # mojolicious controller in case you need
        return $sentence . '?';
    }

To activate this component, you must plug-in this at mojolicious startup method.

    sub startup {
        my $self = shift;
        my $tusu = $self->plugin(tusu => {
            components => {
                'YourUtility' =>  '' ## namespace is ''
            }
        });
    }

Here is another example for component development.

    <div id="productContainer">
        <% Product::list_by_category('books', 10) %>
    </div>

To make it possible, you should write a module like this.

    package Product;
    use strict;
    use warnings;
    use base 'Tusu::ComponentBase';
    
    __PACKAGE__->attr('some_data');
    
    sub init {
        my ($self, $app) = @_;
        $self->some_data('value'); ### DB SETTING OR SOMETHING
    }
    
    sub list_by_category : TplExport {
        my ($self, $category, $limit) = @_;
        my $c = $self->controller; # mojolicious controller in case you need
        
        # MAY BE ACCESS TO YOUR DB HERE
        
        return $html_snippet;
    }

To activate this component, you must plug-in this at mojolicious startup method.

    sub startup {
        my $self = shift;
        my $tusu = $self->plugin(tusu => {
            components => {
                Product => undef
            },
        });
    }

=head1 OPTIONS

=head2 document_root => string

This option sets root directory for templates and static files. Following
example is default setting.

    my $tusu = $self->plugin(tusu => {
        document_root => $self->home->rel_dir('public_html')
    });

=head2 components => hash

    my $tusu = $self->plugin(tusu => {
        components => {
            'Namespace::A' => 'A',   # namespace is A
            'Namespace::B' => '',    # namespace is ''
            'Namespace::C' => undef, # namespace is Namespace::C
        },
    });

=head2 encoding => string or array ref

This option sets encoding for template files. Array ref causes auto detection
active.

    my $tusu = $self->plugin(tusu => {
        encoding => 'Shift-JIS',
    });
    
    or..
    
    my $tusu = $self->plugin(tusu => {
        encoding => ['Shift-JIS', 'utf8'],
    });

=head2 directory_index => array ref

This option sets default file names for searching files in directory when
the request path doesn't ended with file name. And this setting also affects to
inside template context such as include('path') function. Following example is
the default setting.

    my $tusu = $self->plugin(tusu => {
        directory_index => ['index.html', 'index.htm'],
    });

=head2 indexes => bool

This option emulates apache's indexes option. When the value is 1,
the server generates file list page for directory access.

    my $tusu = $self->plugin(tusu => {
        indexes => 1,
    });

=head2 extensions_to_render => array ref

This option sets the extensions to be parsed by tusu renderer. If request
doesn't match any of extensions, dispatcher try to render it as static file.
Following setting is the default.

    my $tusu = $self->plugin(tusu => {
        extensions_to_render => ['html','htm','xml'],
    });

=head2 error_document => hash ref

This option setup custom error pages like apache's ErrorDocument.

    my $tusu = $self->plugin(tusu => {
        error_document => {
            404 => '/errors/404.html',
            403 => '/errors/403.html',
            500 => '/errors/405.html',
        },
    });

=head1 METHODS

=head2 Tusu->new($app)

Constructor. 
    
    $tusu = Tusu->new($app)

=head2 $instance->register($app)

This method internally called.

=head2 $instance->engine

This returns Text::PSTemplate instance. You can customize the template system
behavior by calling parser methods directly.
    
    my $tusu = Tusu->new($app);
    my $pst = $tusu->engine;
    $pst->set_delimiter('<!--', '-->');

=head2 $instance->bootstrap($controller, $component, $method)

This method is a sub dispatcher method. You can specify a class and a method the
route to be dispatched to.

    my $tusu = Tusu->new($self);
    $r->route('/some/path')->via('post')->to(cb => sub {
        $tusu->bootstrap($c, 'Your::Component', 'post');
    });

=head2 $instance->get_component($name)

This is an alias to Text::PSTemplate->get_plugin. With this method, you can get
component instance.

    my $component = $tusu->get_component('Your::Component');

=head1 What does Tusu mean?

Tusu means mojo in Ainu languages which is spoken by the native inhabitants of
Hokkaido prefecture, Japan.

=head1 SEE ALSO

L<Mojolicious>, L<Text::PSTemplate>

L<http://en.wikipedia.org/wiki/Ainu_languages>

=head1 AUTHOR

Sugama Keita, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Sugama Keita.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut
